pkg/provider/vault/cache.go

```diff
@@ -0,0 +1,106 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vault
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru"
+
+	esv1beta1 "github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+)
+
+type clientCache struct {
+	cache       *lru.Cache
+	Size        int
+	initialized bool
+	mu          sync.Mutex
+}
+
+type clientCacheKey struct {
+	Name      string
+	Namespace string
+	Kind      string
+}
+
+type clientCacheValue struct {
+	ResourceVersion string
+	Client          Client
+}
+
+func (c *clientCache) initialize() error {
+	if !c.initialized {
+		var err error
+		c.cache, err = lru.New(c.Size)
+		if err != nil {
+			return fmt.Errorf(errVaultCacheCreate, err)
+		}
+		c.initialized = true
+	}
+	return nil
+}
+
+func (c *clientCache) get(ctx context.Context, store esv1beta1.GenericStore, key clientCacheKey) (Client, bool, error) {
+	value, ok := c.cache.Get(key)
+	if ok {
+		cachedClient := value.(clientCacheValue)
+		if cachedClient.ResourceVersion == store.GetObjectMeta().ResourceVersion {
+			return cachedClient.Client, true, nil
+		}
+		// revoke token and clear old item from cache if resource has been updated
```

Comment:
I found one issue: if we use a static token we must not revoke it.


```go
cachedClient := value.(clientCacheValue)
if cachedClient.ResourceVersion == store.GetObjectMeta().ResourceVersion {
    return cachedClient.Client, true, nil
}
defer c.cache.Remove(key)
// do not revoke if we're having a static token from a secretRef
if store.GetSpec().Provider.Vault.Auth.TokenSecretRef != nil {
    return nil, false, nil
}
// revoke token and clear old item from cache if resource has been updated
err := revokeTokenIfValid(ctx, cachedClient.Client)
if err != nil {
    return nil, false, err
}
```

Reply:
Went for a slightly different approach. Stores using a static token bypass the caching mechanism entirely. No reason to use up memory on these.